core/src/main/java/de/jplag/merging/MatchMerging.java

```diff
@@ -0,0 +1,136 @@
+    private List<JPlagComparison> comparisons;
+    private JPlagOptions options;
+
+    /**
+     * Constructor for class MatchMerging
+     * @param result is the initially computed result object
+     * @param options encapsulates the adjustable options
+     */
+    public MatchMerging(JPlagResult result, JPlagOptions options) {
+        this.result = result;
+        this.comparisons = new ArrayList<>(result.getAllComparisons());
+        this.options = options;
+        this.minimumTokenMatch = options.minimumTokenMatch();
+        this.seperatingThreshold = options.mergingParameters().seperatingThreshold();
+    }
+
+    /**
+     * Runs the internal match merging pipeline. It computes neighboring matches, merges them based on
+     * {@link MergingParameters} and removes the merge buffer afterwards.
+     * @return JPlagResult containing the merged matches
+     */
+    public JPlagResult run() {
+        long timeBeforeStartInMillis = System.currentTimeMillis();
+        for (int i = 0; i < comparisons.size(); i++) {
+            firstSubmission = comparisons.get(i).firstSubmission().copy();
+            secondSubmission = comparisons.get(i).secondSubmission().copy();
+            globalMatches = new ArrayList<>(comparisons.get(i).matches());
+            globalMatches.addAll(comparisons.get(i).ignoredMatches());
+            computeNeighbors();
+            mergeNeighbors();
+            removeBuffer();
+            comparisons.set(i, new JPlagComparison(firstSubmission, secondSubmission, globalMatches, new ArrayList<>()));
+
+        }
+        long durationInMillis = System.currentTimeMillis() - timeBeforeStartInMillis;
+        return new JPlagResult(comparisons, result.getSubmissions(), result.getDuration() + durationInMillis, options);
+    }
+
+    private void computeNeighbors() {
+        neighbors = new ArrayList<>();
+        List<Match> sortedByFirst = new ArrayList<>(globalMatches);
+        Collections.sort(sortedByFirst, (m1, m2) -> m1.startOfFirst() - m2.startOfFirst());
+        List<Match> sortedBySecond = new ArrayList<>(globalMatches);
+        Collections.sort(sortedBySecond, (m1, m2) -> m1.startOfSecond() - m2.startOfSecond());
+        for (int i = 0; i < sortedByFirst.size() - 1; i++) {
+            if (sortedBySecond.indexOf(sortedByFirst.get(i)) == (sortedBySecond.indexOf(sortedByFirst.get(i + 1)) - 1)) {
+                neighbors.add(Arrays.asList(sortedByFirst.get(i), sortedByFirst.get(i + 1)));
+            }
+        }
+    }
+
+    private void mergeNeighbors() {
+        int i = 0;
+        while (i < neighbors.size()) {
+            int lengthUpper = neighbors.get(i).get(0).length();
+            int lengthLower = neighbors.get(i).get(1).length();
+            int seperatingFirst = neighbors.get(i).get(1).startOfFirst() - neighbors.get(i).get(0).endOfFirst() - 1;
+            int seperatingSecond = neighbors.get(i).get(1).startOfSecond() - neighbors.get(i).get(0).endOfSecond() - 1;
```

Comment:
Accessing the neighbors gets really hard to read. Can we redesign this so it is less index magic?
The underlying reason for this is probably the `List<List<Match>>`, which is also not straightforward.
Is it for mapping each match to its neighbors? If yes, we should use a `HashMap<Match, List<Match>`, as this avoids index magic in favor of something like this: `neighbors.get(match);`

Reply:
I have also trouble understanding the magic numbers used here. Why do we only check the first two neighbors? Are there only two because it is the direct ones? Then we should think about appropriate modeling and create a custom datastructure where the semantics are more explicitly modeled.

Reply:
Neighbors contains pairs of neighbouring matches
If there are matches 1,2,3,4 (in Order), neighbours would contain (1,2),(2,3),(3,4)

Reply:
I could use a pair for neighbors intead of an array:
`https://commons.apache.org/proper/commons-lang/javadocs/api-release/org/apache/commons/lang3/tuple/Pair.html`
This would remove one level of index magic, as pairs can be accessed via getLeft() and getRight()